spec/Dynamic-Validator.md

```diff
@@ -83,17 +83,31 @@ It means there can be 100 valid candidates with the potential to be validators.
 The rate of becoming the block proposer is related to the number of delegations that the validator received.
 In other words, CodeChain allows the validator that receives more delegations to generate more blocks than others.
 
+## How to update validators
 ```rust
-let mut validators: Vec<(u64, u64, Account)> = // (Delegation, Deposit, Account)
-let mut proposers: Vec<Account> = vec![];
-validators.reverse_sort();
-let min_delegation = validator.last().0;
-while !validators.is_empty() {
-    for (delegation, _, validator) in validators {
-        proposers.push_back(validator);
-    }
-    validators.retain(|(delegation, _, _)| delegation > min_delegation);
-    validators.for_each(|(mut delegation, _, _)| delegation -= min_delegation);
+let initial_validators: Vec<(Delegation, Deposit, Public)> = // the validator list at the term begin
+let validators: Vec<(Delegation, Deposit, Public)> = // (Delegation, Deposit, Pubkey)
+
+let mut neglecting_validators: Vec<(Delegation, Deposit, Public)> = // ...
+let mut block_author: (Delegation, Deposit, Public) = // block author
+let mut remaining_validators: Vec<(Delegation, Deposit, Public)> = // ...
+assert_eq!(validators, remaining_validators @ block_author :: neglecting_validators);
+
+let min_delegation: u64 = initial_validators.into_iter().map(|(delegation, _, _)| delegation).min();
+
+block_author.0 = block_author.0.saturating_sub(min_delegation);
+remaining_validators.push(block_author);
+
+for neglecting in neglecting_validators {
+    // To prevent the slow block generation when some nodes stopped.
+    neglecting_validators.0 = neglecting_validators.0.saturating_sub(min_delegation * 2);
+}
+remaining_validators.append(&mut neglecting_validators);
+
+remaining_validators.sort();
+
+if remaining_validators.into_iter().all(|(delegation, _, _)| delegation == 0) {
+    remaining_validators = initial_validators;
```

Comment:
I think this psuedo-code would be much readable if we rewrite it to this
```rust
let initial_validators: Vec<(Delegation, Deposit, Public)> = // the validator list at the term begin
let mut validators: Vec<(Delegation, Deposit, Public)> = // (Delegation, Deposit, Pubkey)

let author_index = validators.position(|(_, _, pubkey)| pubkey == block_author).unwrap();
for (mut ref delegation, _, pubkey) in validators[0..author_index] {
    // Deprioritize fast since they are neglicting as a validator.
	delegation -= min_delegation * 2;
}
// Deprioritize author
validators[author_index].2 -= min_delegation;

validators.sort();

if validators.into_iter().all(|(delegation, _, _)| delegation == 0) {
    validators = initial_validators;
}
```

Reply:
It's
```rust
for (mut ref delegation, _, pubkey) in validators[author_index..] {
```
Anyway, I agree your comment. I'll update it.